### PR TITLE
Dispose of most uses of LogicalType and Node

### PR DESCRIPTION
**Pull request by adamreeve**
In this PR I've tried to dispose of most internal uses of LogicalType and Node to reduce the need to run finalizers.

There is still one main place where we create logical types and nodes that aren't disposed that I'm aware of. In LogicalColumnStream, the LogicalType and SchemaNodesPath properties are initialized in the constructor rather than being getters that create new instances, so these should probably be disposed when disposing the LogicalColumnStream instance. But there's a small chance that could be a breaking change if somebody is expecting to be able to continue using those after the LogicalColumnStream is disposed, so I've currently left that as-is.

There's also `RowOriented.ParquetFile.GetColumn` which creates logical types that aren't disposed, and I couldn't see a straightforward way to fix that.

### CODE REVIEW

File: csharp/ParquetFileReader.cs [hunk 1]
```diff
@@ -7,33 +7,37 @@ namespace ParquetSharp
...
         }
 
         public ParquetFileReader(RandomAccessFile randomAccessFile)
-            : this(randomAccessFile, ReaderProperties.GetDefaultReaderProperties())
+            : this(randomAccessFile, null)
         {
         }
 
-        public ParquetFileReader(string path, ReaderProperties readerProperties)
+        public ParquetFileReader(string path, ReaderProperties? readerProperties)
```
Comment (adamreeve) [inline csharp/ParquetFileReader.cs:19]: Nullability of reference types isn't part of the binary method signature so this is ABI compatible as well as source compatible.

File: csharp/ColumnDescriptor.cs [hunk 2]
```diff
@@ -69,11 +69,22 @@ public TReturn Apply<TReturn>(LogicalTypeFactory typeFactory, Type? columnLogica
             var (physicalType, logicalType) = typeFactory.GetSystemTypes(this, columnLogicalTypeOverride);
             var elementType = logicalType;
 
-            for (var node = SchemaNode; node != null; node = node.Parent)
+            var node = SchemaNode;
+            while (node != null)
             {
-                if (node.LogicalType.Type == LogicalTypeEnum.List)
+                try
```
Comment (marcin-krystianc) [inline csharp/ColumnDescriptor.cs:75]: I don't think that `try/finally` improves anything here. In my opinion it just adds complexity to the code without any real benefit.

File: csharp/Schema/GroupNode.cs [hunk 3]
```diff
@@ -37,11 +37,24 @@ public int FieldIndex(Node node)
...
-                LogicalType is NoneLogicalType ? null : LogicalType);
+            using var logicalType = LogicalType;
+            var fields = Fields;
+            var clonedFields = fields.Select(f => f.DeepClone()).ToArray();
+            try
+            {
+                return new GroupNode(
+                    Name,
+                    Repetition,
+                    fields,
```
Comment (marcin-krystianc) [inline csharp/Schema/GroupNode.cs:48]:
```suggestion
                    clonedFields,
```

File: csharp/Schema/GroupNode.cs [hunk 4]
```diff
@@ -37,11 +37,24 @@ public int FieldIndex(Node node)
...
+                    logicalType is NoneLogicalType ? null : logicalType);
+            }
+            finally
+            {
+                foreach (var field in fields.Concat(clonedFields))
```
Comment (marcin-krystianc) [inline csharp/Schema/GroupNode.cs:53]: Is it correct to dispose `clonedFields` here ?

File: csharp.test/TestByteArrayReaderCache.cs [hunk 5]
```diff
@@ -29,7 +29,8 @@ public static void TestReadingDuplicateStrings([Values(true, false)] bool enable
             // Write a file that contains a lot of duplicate strings.
             using (var output = new BufferOutputStream(buffer))
             {
-                using var fileWriter = new ParquetFileWriter(output, columns, CreateWriterProperties(enableDictionary));
+                using var writerProperties = CreateWriterProperties(enableDictionary);
```
[the remark this answers is not in the document]
Comment (adamreeve) [reply csharp.test/TestByteArrayReaderCache.cs:32]: Yeah this pattern is used a lot throughout the tests when writing and then needing to dispose the writing related objects before reading, so I wouldn't want to change that here. It also doesn't bother me too much, I think it's fairly obvious in both cases when the different objects will be disposed.

File: csharp/ColumnDescriptor.cs (hunk 2 above)
Comment (adamreeve) [reply csharp/ColumnDescriptor.cs:75]: So you mean just remove the try/finally and keep the rest of the code as-is, so if there's an exception we let the finalizers run to free the handles? That seems reasonable to me, especially as this code is highly unlikely to throw.

File: csharp/Schema/GroupNode.cs (hunk 3 above)
Comment (adamreeve) [reply csharp/Schema/GroupNode.cs:48]: A whoops, good catch thanks.

File: csharp/Schema/GroupNode.cs (hunk 4 above)
Comment (adamreeve) [reply csharp/Schema/GroupNode.cs:53]: Yes, once we've created the GroupNode, it no longer references the original C# side fields but internally in the C++ GroupNode it has shared pointers that point to the fields. Disposing of the field just deletes the shared pointer to the C++ field, but the field itself won't be deleted due to the group node still holding a shared pointer to it.

File: csharp/ColumnDescriptor.cs (hunk 2 above)
Comment (adamreeve) [reply csharp/ColumnDescriptor.cs:75]: Do you think I should similarly remove `try/finally` elsewhere, eg. in the tests?

File: csharp/Schema/GroupNode.cs (hunk 4 above)
Comment (marcin-krystianc) [reply csharp/Schema/GroupNode.cs:53]: 👍

File: csharp/ColumnDescriptor.cs (hunk 2 above)
Comment (adamreeve) [reply csharp/ColumnDescriptor.cs]: Right, good point, I've changed this now.

**Review by marcin-krystianc — APPROVED** [review]

**Pull request merged**